STAR/draft-ietf-acme-star.md

```diff
@@ -368,45 +370,76 @@ Request" and type "malformedRequest".
 
 - star-certificate (optional, string):  A URL for the (rolling) STAR certificate that has been issued in response to this Order.
 
+
 ### Canceling an Auto-renewal Order
 {: #protocol-details-canceling}
 
 An important property of the auto-renewal Order is that it can be canceled by the IdO, with no need for certificate revocation. To cancel the Order, the ACME client sends a POST to the Order URL as shown in {{figcancelingstarorder}}.
 
 ~~~
-  POST /acme/order/TOlocE8rfgo HTTP/1.1
+  POST /acme/order/ogfr8EcolOT HTTP/1.1
   Host: example.org
   Content-Type: application/jose+json
 
   {
     "protected": base64url({
       "alg": "ES256",
-      "kid": "https://example.com/acme/acct/evOfKhNU60wg",
-      "nonce": "5XJ1L3lEkMG7tR6pA00clA",
-      "url": "https://example.com/acme/order/TOlocE8rfgo"
+      "kid": "https://example.com/acme/acct/gw06UNhKfOve",
+      "nonce": "Alc00Ap6Rt7GMkEl3L1JX5",
+      "url": "https://example.com/acme/order/ogfr8EcolOT"
     }),
     "payload": base64url({
       "status": "canceled"
     }),
-    "signature": "H6ZXtGjTZyUnPeKn...wEA4TklBdh3e454g"
+    "signature": "g454e3hdBlkT4AEw...nKePnUyZTjGtXZ6H"
   }
 ~~~
 {: #figcancelingstarorder title="Canceling an Auto-renewal Order"}
 
 After a successful cancellation, the server MUST NOT issue any additional certificates for this Order.
 
-Immediately after the Order is canceled, the server:
+When the Order is canceled, the server:
 
 - MUST update the status of the Order resource to "canceled" and MUST set an appropriate "expires" date;
 - MUST respond with 403 (Forbidden) to any requests to the star-certificate endpoint.  The response SHOULD provide
 additional information using a problem document {{RFC7807}} with type "urn:ietf:params:acme:error:autoRenewalCanceled".
 
 Issuing a cancellation for an Order that is not in "valid" state is not allowed.  A client MUST NOT send such a request, and a server MUST return an error response with status code 400 (Bad Request) and type "urn:ietf:params:acme:error:autoRenewalCancellationInvalid".
 
+The state machine described in Section 7.1.6 or {{RFC8555}} is extended as illustrated in {{fig-order-state-transitions-ext}}.
+
+~~~
+    pending --------------+
+       |                  |
+       | All authz        |
+       | "valid"          |
+       V                  |
+     ready ---------------+
+       |                  |
+       | Receive          |
+       | finalize         |
+       | request          |
+       V                  |
+   processing ------------+
+       |                  |
+       | Certificate      | Error or
+       | issued           | Authorization failure
+       V                  V
+     valid             invalid
+       |
+       | STAR
+       | Certificate
+       | canceled
+       V
+    canceled
+
```

Comment:
I think "First certificate issued" is correct -- i.e., we don't need further clarifications.  In fact, somewhere else we say: `the status MUST be "valid" as long as STAR certificates are being issued.`

Reply:
I think the state moves to "valid" even earlier. Think of the case where `start-date` is specified, and it's a week after authorization. I think we would move to "valid" immediately without waiting for the first cert to be issued.

Reply:
Makes sense.

Reply:
I agree about the drift, but I think it does get us the ability to cancel an Order before the first cert has been issued, which is valuable.

Reply:
Having said that, the whole cancellation thing is a STAR artifact, so we could allow it for both Valid and Processing states, and then keep the Valid state as-is.

Reply:
> I think it does get us the ability to cancel an Order before the first cert has been issued, which is valuable.

Maybe here's the disconnect: I thought the star-cert resource is populated as soon as the order is finalised (possibly with a future notBefore, depending on auto-renewal.start-date)?  I don't see a reason why the first STAR cert should stay in limbo after the authorisation phase completes, but maybe I'm missing something...

Reply:
Got it. But first, we are not clear on this either way - if we decide that the first cert is issued immediately, we should say so in the "Fetching the Certificates" section. And second, I think there's significant benefit in being able to cancel the Order in the time period before the first cert's validity, a period which may be measured in weeks (it will be a manual process, most likely). The alternative is, if a breach is detected during that time, you are unable to cancel and the first cert is like a time bomb waiting to go off.

Reply:
> Got it. But first, we are not clear on this either way - if we decide that the first cert is issued immediately, we should say so in the "Fetching the Certificates" section.

OK, I haven't re-read that, but if you think clarifying that section would fix all current ambiguities, let's do it!

> And second, I think there's significant benefit in being able to cancel the Order in the time period before the first cert's validity, a period which may be measured in weeks (it will be a manual process, most likely). The alternative is, if a breach is detected during that time, you are unable to cancel and the first cert is like a time bomb waiting to go off.

I fully agree, but I think this would not even be a problem if the cert goes to "valid" -- although, maybe, in a lethargic state -- just after order's finalisation (as per base ACME).  Because this makes cancellation immediately available to the client.

Reply:
But then the arrow should say something like "Order finalized" rather than "first cert issued". Because we leave it open whether the cert is immediately issued or not. I'm not very happy with this solution because if someone steals the cert off the star-certificate URL, you will never know about it and therefore you don't know if your cancellation was effective.

I strongly prefer an implementation where the CA only issues the cert when it needs to, and not earlier.

Reply:
I can live with that. Let's move forward.

Reply:
OK; I've added a description of the attack and mitigation to the "No revocation" subsection of the security considerations in fa792555